async-http-client-backend/zio/src/test/scala/sttp/client/asynchttpclient/zio/WebSocketStubZioTests.scala

```diff
@@ -0,0 +1,117 @@
+
+import scala.util.{ Failure, Success }
+
+class WebSocketStubZioTests extends AnyFlatSpec with Matchers with ScalaFutures {
+
+  def mkBackend(webSocketStub: WebSocketStub[_]) =
+    AsyncHttpClientZioBackend.stub
+      .whenRequestMatches(_ => true)
+      .thenRespondWebSocket(Headers(List.empty), webSocketStub)
+
+  "web socket stub" should "return initial Incoming frames on 'receive'" in {
+    val frames = List("a", "b", "c").map(WebSocketFrame.text(_))
+    val webSocketStub = WebSocketStub.withInitialIncoming(frames)
+    implicit val b = mkBackend(webSocketStub)
+    val test = for {
+      handler <- ZioWebSocketHandler()
+      ws <- basicRequest.get(uri"http://example.org/a/b/c").openWebsocket(handler).map(_.result)
+      msg1 <- ws.receive
+      msg2 <- ws.receive
+      msg3 <- ws.receive
+    } yield List(msg1, msg2, msg3)
+
+    runtime.unsafeRun(test) shouldBe frames.map(Right(_))
+  }
+
+  it should "return initial responses on 'receive'" in {
+    val okFrame = WebSocketFrame.text("abc")
+    val exception = new Exception("boom")
+    val webSocketStub = WebSocketStub.withInitialMessages(List(Success(Right(okFrame)), Failure(exception)))
+    implicit val b = mkBackend(webSocketStub)
+    val test = for {
+      handler <- ZioWebSocketHandler()
+      ws <- basicRequest.get(uri"http://example.org/a/b/c").openWebsocket(handler).map(_.result)
+      msg <- ws.receive
+      err <- ws.receive.either
+    } yield (msg, err)
+
+    runtime.unsafeRun(test) shouldBe ((Right(okFrame), Left(exception)))
+  }
+
+  it should "add next Incoming frames as reaction to send" in {
+    val firstFrame = WebSocketFrame.text("No. 1")
+    val secondFrame = WebSocketFrame.text("Second")
+    val thridFrame = WebSocketFrame.text("3")
+    val expectedFrame = WebSocketFrame.text("give me more!")
+    val webSocketStub = WebSocketStub
+      .withInitialIncoming(List(firstFrame))
+      .thenRespond {
+        case `expectedFrame` => List(secondFrame, thridFrame)
+        case _               => List.empty
+      }
+    implicit val b = mkBackend(webSocketStub)
+    val test = for {
+      handler <- ZioWebSocketHandler()
+      ws <- basicRequest.get(uri"http://example.org/a/b/c").openWebsocket(handler).map(_.result)
+      msg1 <- ws.receive
+      err1 <- ws.receive.either
+      _ <- ws.send(WebSocketFrame.text("not expected"))
+      err2 <- ws.receive.either
+      _ <- ws.send(expectedFrame)
+      msg2 <- ws.receive
+      msg3 <- ws.receive
+    } yield (msg1, err1.isLeft, err2.isLeft, msg2, msg3)
+
+    runtime.unsafeRun(test) shouldBe ((Right(firstFrame), true, true, Right(secondFrame), Right(thridFrame)))
+  }
+
+  it should "add next responses as reaction to send" in {
+    val exception = new Exception("transient error")
+    val ok = WebSocketFrame.text("ok")
+
+    val webSocketStub = WebSocketStub.withNoInitialMessages
+      .thenRespondWith {
+        case _ => List(Failure(exception), Success(Right(ok)))
```

Comment:
shouldn't the websocket get closed after the exception or a `Close` event? That is, a websocket which already produced an exception or a close should be unusable later on

Reply:
So, there is no such thing as transient error in WebSocket?

Reply:
not that I'm aware of, if the TCP conneciton gets broken, it's broken, I wouldn't use it again ;)

Reply:
Then I'll make WebSocketStub stay in last failure and also fail after it returned closed.